### PR TITLE
feat/gen-metrics-counter-schema

### PR DESCRIPTION
**Pull request by nikhars**
### Context
The generic metrics platform does not have the counters dataset. It wasn't created because there was no need for the dataset. Now there is a need for counters on the generic metrics platform.

### Additional information
The dataset is very similar to `generic_metrics_sets` or `generic_metrics_distributions`. The primary difference is how it populates the count.

### Testing
Validated that the migration was run successfully and the tables were created

generic_metrics
[X]  0001_sets_aggregate_table
[X]  0002_sets_raw_table
[X]  0003_sets_mv
[X]  0004_sets_raw_add_granularities
[X]  0005_sets_replace_mv
[X]  0006_sets_raw_add_granularities_dist_table
[X]  0007_distributions_aggregate_table
[X]  0008_distributions_raw_table
[X]  0009_distributions_mv
[X]  0010_counters_aggregate_table
[X]  0011_counters_raw_table
[X]  0012_counters_mv

6749ecd45d3d :) show create table generic_metric_counters_aggregated_local

SHOW CREATE TABLE generic_metric_counters_aggregated_local

┌─statement──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE default.generic_metric_counters_aggregated_local (`org_id` UInt64, `project_id` UInt64, `metric_id` UInt64, `granularity` UInt8, `timestamp` DateTime CODEC(DoubleDelta), `retention_days` UInt16, `tags.key` Array(UInt64), `tags.indexed_value` Array(UInt64), `tags.raw_value` Array(String), `value` AggregateFunction(sum, Float64), `use_case_id` LowCardinality(String), `_indexed_tags_hash` Array(UInt64) MATERIALIZED arrayMap((k, v) -> cityHash64(concat(toString(k), '=', toString(v))), tags.key, tags.indexed_value), `_raw_tags_hash` Array(UInt64) MATERIALIZED arrayMap((k, v) -> cityHash64(concat(toString(k), '=', v)), tags.key, tags.raw_value),  INDEX bf_indexed_tags_hash _indexed_tags_hash TYPE bloom_filter() GRANULARITY 1,  INDEX bf_raw_tags_hash _raw_tags_hash TYPE bloom_filter() GRANULARITY 1,  INDEX bf_tags_key_hash tags.key TYPE bloom_filter() GRANULARITY 1) ENGINE = AggregatingMergeTree() PARTITION BY (retention_days, toMonday(timestamp)) PRIMARY KEY (org_id, project_id, metric_id, granularity, timestamp) ORDER BY (org_id, project_id, metric_id, granularity, timestamp, tags.key, tags.indexed_value, tags.raw_value, retention_days, use_case_id) TTL timestamp + toIntervalDay(retention_days) SETTINGS index_granularity = 2048 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 rows in set. Elapsed: 0.006 sec.

6749ecd45d3d :) show create table generic_metric_counters_raw_local

SHOW CREATE TABLE generic_metric_counters_raw_local

┌─statement──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE default.generic_metric_counters_raw_local (`use_case_id` LowCardinality(String), `org_id` UInt64, `project_id` UInt64, `metric_id` UInt64, `timestamp` DateTime, `retention_days` UInt16, `tags.key` Array(UInt64), `tags.indexed_value` Array(UInt64), `tags.raw_value` Array(String), `set_values` Array(UInt64), `count_value` Float64, `distribution_values` Array(Float64), `metric_type` LowCardinality(String), `materialization_version` UInt8, `timeseries_id` UInt32, `partition` UInt16, `offset` UInt64, `granularities` Array(UInt8)) ENGINE = MergeTree() PARTITION BY toStartOfInterval(timestamp, toIntervalDay(3)) ORDER BY (use_case_id, org_id, project_id, metric_id, timestamp) TTL timestamp + toIntervalDay(7) SETTINGS index_granularity = 8192 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 rows in set. Elapsed: 0.005 sec.

6749ecd45d3d :) show create table generic_metric_counters_aggregation_mv

SHOW CREATE TABLE generic_metric_counters_aggregation_mv

┌─statement──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE MATERIALIZED VIEW default.generic_metric_counters_aggregation_mv TO default.generic_metric_counters_aggregated_local (`org_id` UInt64, `project_id` UInt64, `metric_id` UInt64, `granularity` UInt8, `timestamp` DateTime CODEC(DoubleDelta), `retention_days` UInt16, `tags.key` Array(UInt64), `tags.indexed_value` Array(UInt64), `tags.raw_value` Array(String), `value` AggregateFunction(sum, Float64), `use_case_id` LowCardinality(String)) AS SELECT use_case_id, org_id, project_id, metric_id, arrayJoin(granularities) AS granularity, tags.key, tags.indexed_value, tags.raw_value, toDateTime(multiIf(granularity = 0, 10, granularity = 1, 60, granularity = 2, 3600, granularity = 3, 86400, -1) * intDiv(toUnixTimestamp(timestamp), multiIf(granularity = 0, 10, granularity = 1, 60, granularity = 2, 3600, granularity = 3, 86400, -1))) AS timestamp, retention_days, sumState(count_value) AS value FROM default.generic_metric_counters_raw_local WHERE (materialization_version = 1) AND (metric_type = 'counter') GROUP BY use_case_id, org_id, project_id, metric_id, tags.key, tags.indexed_value, tags.raw_value, timestamp, granularity, retention_days │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 rows in set. Elapsed: 0.003 sec.

### CODE REVIEW

**Review by onewland — APPROVED** [review]
This mostly looks good to me, except the storage set key. I do remember @fpacifici saying at one point that I had constructed indexes wrong on some table[s], so we should make sure that mistake isn't repeated here

File: snuba/snuba_migrations/generic_metrics/0011_counters_raw_table.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.migrations.operations import OperationTarget
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+    local_table_name = "generic_metric_counters_raw_local"
+    dist_table_name = "generic_metric_counters_raw_dist"
+    storage_set_key = StorageSetKey.GENERIC_METRICS_DISTRIBUTIONS
```
Comment (onewland) [inline snuba/snuba_migrations/generic_metrics/0011_counters_raw_table.py:22]: wrong storage set key4

File: snuba/snuba_migrations/generic_metrics/0010_counters_aggregate_table.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        Column("value", AggregateFunction("sum", [Float(64)])),
+        Column("use_case_id", String(Modifiers(low_cardinality=True))),
+    ]
+
+    def forwards_ops(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=self.storage_set_key,
+                table_name=self.local_table_name,
+                engine=table_engines.AggregatingMergeTree(
```
Comment (fpacifici) [inline snuba/snuba_migrations/generic_metrics/0010_counters_aggregate_table.py:55]: Have you considered using a summing megetree here instead ? And what would be the pros and cons ?

Comment (nikhars) [reply snuba/snuba_migrations/generic_metrics/0010_counters_aggregate_table.py:55]: I haven't considered using the SummingMergeTree. I used the AggregatingMergeTree based on what is being done by the existing counters table of release health metrics dataset. 

Let me know if you have a preference on which one to use (with possibly the reasoning behind it). I will see at that point if I need to change something or not

File: snuba/snuba_migrations/generic_metrics/0011_counters_raw_table.py (hunk 1 above)
Comment (nikhars) [reply snuba/snuba_migrations/generic_metrics/0011_counters_raw_table.py]: Good find. Fixed it.

**Review by onewland — APPROVED** [review]
new changes look good to me. I do think the test failures are legitimate, though

**Pull request merged**